File: lp/trab3/livro.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-

###
#   livro.py
#
#   Definição e implementação da
#   classe Livro e suas funções
###

from datetime import date

class Livro:
    def __init__(self, codigo, titulo, autor, assunto, dataPub, editora, resumo):
        self.__codigo = codigo
        self.__titulo = titulo
        self.__autor = autor
        self.__assunto = assunto
        self.__dataPub = dataPub
        self.__editora = editora
        self.__resumo = resumo

    # Retorna o código do livro
    def getCodigo(self):
        return self.__codigo

    # Retorna o título do livro
    def getTitulo(self):
        return self.__titulo

    # Retorna o autor do livro
    def getAutor(self):
        return self.__autor

    # Retorna a data de publicação do livro
    def getData(self):
        return self.__dataPub

    # Retorna os dados do livro em uma lista
    def getDados(self):
        dados = [self.__codigo] + [self.__titulo] + [self.__autor] + [self.__assunto] + [self.__dataPub.strftime("%d/%m/%Y")] + [self.__editora] + [self.__resumo]
        return dados

    @staticmethod #Compara crescentemente por Código
    def comparaCodigo(livro1, livro2):
        cod1 = int(livro1.getCodigo())
        cod2 = int(livro2.getCodigo())

        if cod1 < cod2:
            return -1
        elif cod1 == cod2:
            return 0
        else:
            return 1

    @staticmethod #Compara decrescentemente por Título
    def comparaTitulo(livro1, livro2):
        titulo1 = livro1.getTitulo()
        titulo2 = livro2.getTitulo()

        if titulo1 > titulo2:
            return -1
        elif titulo1 == titulo2:
            return (-1)*Livro.comparaCodigo(livro1, livro2)
        else:
            return 1

    @staticmethod #Compara crescentemente por Autor
    def comparaAutor(livro1, livro2):
        autor1 = livro1.getAutor()
        autor2 = livro2.getAutor()

        if autor1 < autor2:
            return -1
        elif autor1 == autor2:
            return Livro.comparaCodigo(livro1, livro2)
        else:
            return 1

    @staticmethod #Compara decrescentemente por Data
    def comparaData(livro1, livro2):
        data1 = livro1.getData()
        data2 = livro2.getData()

        if data1 > data2:
            return -1
        elif data1 == data2:
            return (-1)*Livro.comparaCodigo(livro1, livro2)
        else:
            return 1

File: lp/trab3/Livro.hs
--
-- Livro.hs
--
-- Definição e implementação da
-- classe Livro e suas funções
--

-- Definição do módulo Livro
module Livro
( Livro(..)
, comparaLivroPorCodigo
, comparaLivroPorTitulo
, comparaLivroPorAutor
, comparaLivroPorData
) where

-- Importando Data.hs
import Data

-- Definição da estrutura Livro
data Livro  = Livro {
    codigo  :: String,
    titulo  :: String,
    autor   :: String,
    assunto :: String,
    dataPub :: Data,
    editora :: String,
    resumo  :: String
}

-- Como mostrar um Livro
instance Show Livro where
    show (Livro c t a ass d e r) = id c ++ "\n" ++ id t ++ "\n" ++ id a ++ "\n" ++ id ass ++ "\n" ++ show d ++ "\n" ++ id e ++ "\n" ++ id r

-- Como comparar um Livro...
-- 1. por código (crescente):
comparaLivroPorCodigo l1 l2
    | cod1 <= cod2  = True
    | otherwise     = False

    where cod1 = read (codigo l1) :: Int
          cod2 = read (codigo l2) :: Int

-- 2. por título (decrescente):
comparaLivroPorTitulo l1 l2
    | titulo l1 > titulo l2   = True
    | titulo l1 == titulo l2  = not (comparaLivroPorCodigo l1 l2)
    | otherwise               = False

-- 3. por autor (crescente):
comparaLivroPorAutor l1 l2
    | autor l1 < autor l2   = True
    | autor l1 == autor l2  = comparaLivroPorCodigo l1 l2
    | otherwise             = False

-- 3. por data (decrescente):
comparaLivroPorData l1 l2
    | dataPub l1 > dataPub l2   = True
    | dataPub l1 == dataPub l2  = not (comparaLivroPorCodigo l1 l2)
    | otherwise                 = False
